File: DAIRemoteApp/app/src/main/java/com/example/dairemote_app/ConnectionManager.java
package com.example.dairemote_app;

import android.os.Build;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public class ConnectionManager {
    private final ExecutorService executorService;
    private static ExecutorService hostSearchExecService;
    private static final AtomicBoolean connectionEstablished = new AtomicBoolean(false);
    private static String serverAddress;
    private String hostName;
    private String hostAudioList;
    private String hostDisplayProfileList;
    private String hostRequesterResponse;
    private static String serverResponse;
    private static InetAddress inetAddress;
    private static final int serverPort = 11000;
    private static HostSearchCallback hostHandler;

    public static byte[] receiveData = new byte[1024];
    private static byte[] sendData = new byte[200];
    private static DatagramPacket sendPacket;
    public static DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
    private static DatagramSocket udpSocket;

    static {
        try {
            udpSocket = new DatagramSocket();
        } catch (SocketException e) {
            throw new RuntimeException(e);
        }
    }

    private static final InetAddress broadcastAddress;

    static {
        try {
            broadcastAddress = InetAddress.getByName("255.255.255.255");
        } catch (UnknownHostException e) {
            throw new RuntimeException(e);
        }
    }

    public ConnectionManager(String serverAddress) {
        try {
            SetServerAddress(serverAddress);
            SetServerAddress(InetAddress.getByName(GetServerAddress()));
        } catch (Exception ignored) {
        }

        this.executorService = Executors.newCachedThreadPool();
    }

    public static DatagramSocket GetUDPSocket() {
        return udpSocket;
    }

    public static void SetServerAddress(String address) {
        serverAddress = address;
    }

    public static void SetServerAddress(InetAddress address) {
        inetAddress = address;
    }

    public static String GetServerAddress() {
        return serverAddress;
    }

    public static InetAddress GetInetAddress() {
        return inetAddress;
    }

    public static int GetPort() {
        return serverPort;
    }

    public void SetConnectionEstablished(boolean status) {
        connectionEstablished.set(status);
    }

    public static boolean GetConnectionEstablished() {
        return connectionEstablished.get();
    }

    private static void SetServerResponse(String response) {
        serverResponse = response;
    }

    public static String GetServerResponse() {
        return serverResponse;
    }

    public void SetHostAudioList(String list) {
        this.hostAudioList = list;
    }

    public String GetHostAudioList() {
        return this.hostAudioList;
    }

    public void SetHostDisplayProfilesList(String list) {
        this.hostDisplayProfileList = list;
    }

    public String GetHostDisplayProfilesList() {
        return this.hostDisplayProfileList;
    }

    public void SetHostRequesterResponse(String response) {
        this.hostRequesterResponse = response;
    }

    public String GetHostRequesterResponse() {
        return this.hostRequesterResponse;
    }

    public void SetHostName(String name) {
        this.hostName = name;
    }

    public String GetHostName() {
        if(this.hostName != null) {
            return this.hostName.substring("HostName: ".length());
        }
        return "";
    }

    public static void SendData(String message, InetAddress address) throws IOException {
        if (udpSocket == null || udpSocket.isClosed()) {
            udpSocket = new DatagramSocket();
        }
        sendData = message.getBytes();
        sendPacket = new DatagramPacket(sendData, sendData.length, address, GetPort());
        udpSocket.send(sendPacket);
    }

    public static void ReceiveData() throws IOException {
        // Clear prior data
        Arrays.fill(receiveData, (byte) 0);
        udpSocket.receive(receivePacket);
    }

    // Utility to get the device name
    public static String GetDeviceName() {
        String manufacturer = Build.MANUFACTURER;
        String model = Build.MODEL;
        if (model.startsWith(manufacturer)) {
            return model;
        } else {
            return manufacturer + " " + model;
        }
    }

    // Broadcast to search for hosts in the background
    public static void HostSearchInBackground(HostSearchCallback hostSearchCallback, String message) {
        hostHandler = hostSearchCallback;
        hostSearchExecService = Executors.newSingleThreadExecutor();
        // Perform the host search
        hostSearchExecService.execute(() -> HostSearch(message));
    }

    public static void ShutdownHostSearchInBackground() {
        hostSearchExecService.shutdownNow();
    }

    // Broadcast to search for hosts
    public static void HostSearch(String message) {
        List<String> hosts = new ArrayList<>();
        try {
            if(udpSocket == null || udpSocket.isClosed()) {
                udpSocket = new DatagramSocket();
            }
            udpSocket.setBroadcast(true);
            SendData(message + " " + GetDeviceName(), broadcastAddress);
            udpSocket.setSoTimeout(3000); // Millisecond timeout for responses and to break out of loop
            while (true) {
                try {

                    ReceiveData(); // Blocks until a response is received

                    SetServerResponse(new String(receivePacket.getData()).trim());
                    String serverIp = receivePacket.getAddress().getHostAddress();

                    if (GetServerResponse().contains("Hello, I'm")) {
                        hosts.add(serverIp);
                        udpSocket.setSoTimeout(75);    // Reset timeout on host found, otherwise lingers
                        udpSocket.setBroadcast(false);
                    }
                } catch (SocketTimeoutException e) {
                    // Stop listening for responses, occurs on timeout
                    break;
                }
            }

            // Reset server response, just being used for broadcast search here
            // Needs to be empty for proceeding functions
            SetServerResponse(null);

            if (!hosts.isEmpty()) {
                hostHandler.onHostFound(hosts);
            } else {
                hostHandler.onError("No hosts found");
            }
        } catch (SocketException ignored) {
        } catch (UnknownHostException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    // Wait for server response
    public void WaitForResponse(int timeout) {
        try {
            udpSocket.setSoTimeout(timeout);
            ReceiveData();
            udpSocket.setSoTimeout(75);

            SetServerResponse(new String(receivePacket.getData(), 0, receivePacket.getLength()));
        } catch (Exception ignored) {
            SetServerResponse("");
        }
    }

    // Initialize connection
    public boolean InitializeConnection() {
        SetServerResponse("");
        int broadcastCount = 0;
        while (GetServerResponse().isEmpty()) {
            try {
                SendData("Connection requested by " + GetDeviceName(), GetInetAddress());
            } catch (Exception ignored) {
            }
            broadcastCount += 1;
            if (broadcastCount > 5) {
                return false;
            } else {
                // Updates serverResponse else times out and throws socket exception
                try {
                    WaitForResponse(5000);
                } catch (Exception ignored) {
                }
            }

        }
        return FinishConnection();
    }

    public boolean FinishConnection() {
        int approvalTimeout = 0;
        if (GetServerResponse().equalsIgnoreCase("Wait")) {
            SetServerResponse("");
            while (GetServerResponse().isEmpty()) {
                WaitForResponse(10000);
                approvalTimeout += 1;
                if(approvalTimeout > 5) {
                    break;
                }
            }
            if (GetServerResponse().equalsIgnoreCase("Approved")) {
                ShutdownHostSearchInBackground();
                ConnectionMonitor.GetInstance(MainActivity.connectionManager);
                SetConnectionEstablished(true);
                return true;
            } else if (GetServerResponse().equalsIgnoreCase("Connection attempt declined.")) {
                ResetConnectionManager();
            }
            SetServerResponse("");
        } else if (GetServerResponse().equalsIgnoreCase("Approved")) {
            ShutdownHostSearchInBackground();
            ConnectionMonitor.GetInstance(MainActivity.connectionManager);
            SetConnectionEstablished(true);
            return true;
        }
        return false;
    }

    public void StopExecServices(ExecutorService service) {
        if (service != null && !service.isShutdown()) {
            service.shutdownNow();
        }
    }

    // Send message to the server
    public boolean SendHostMessage(String msg) {
        if (ConnectionManager.connectionEstablished.get()) {
            executorService.submit(() -> SendMessage(msg));
            return true;
        }
        return false;
    }

    public void SendMessage(String message) {
        try {
            SendData(message, GetInetAddress());
        } catch (Exception ignored) {
        }
    }

    public boolean HostRequester(String replyCondition, String sendMessage, InetAddress inetAddress) {
        SetServerResponse("");
        int broadcastCount = 0;
        while (!GetServerResponse().startsWith(replyCondition)) {
            try {
                SendData(sendMessage, inetAddress);
            } catch (Exception ignored) {
            }
            broadcastCount += 1;
            if (broadcastCount > 5) {
                return false;
            } else {
                // Updates serverResponse else times out and throws socket exception
                try {
                    WaitForResponse(5000);
                } catch (Exception ignored) {
                }
            }

        }
        SetHostRequesterResponse(GetServerResponse());
        return true;
    }

    public boolean RequestHostName() {
        if(HostRequester("HostName", "HOST Name", GetInetAddress())) {
            SetHostName(GetHostRequesterResponse());
            return true;
        }
        return false;
    }

    // Retrieve audio devices from host
    public boolean RequestHostAudioDevices() {
        if(HostRequester("AudioDevices", "AUDIO Devices", GetInetAddress())) {
            SetHostAudioList(GetHostRequesterResponse());
            return true;
        }
        return false;
    }

    // Retrieve display profiles from host
    public boolean RequestHostDisplayProfiles() {
        if(HostRequester("DisplayProfiles", "DISPLAY Profiles", GetInetAddress())) {
            SetHostDisplayProfilesList(GetHostRequesterResponse());
            return true;
        }
        return false;
    }

    public void ResetConnectionManager() {
        StopExecServices(executorService);
        SetConnectionEstablished(false);
        SetServerResponse(null);
        SetServerAddress((String) null);
        SetServerAddress((InetAddress) null);
    }

    // Shutdown the connection
    public void Shutdown() {
        ConnectionMonitor.GetInstance(this).StopHeartbeat();
        SendHostMessage("Shutdown requested");
        try {
            Thread.sleep(75);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        ResetConnectionManager();
    }

    public void CloseUDPSocket() {
        if (udpSocket != null && !udpSocket.isClosed()) {
            udpSocket.close();
        }
    }
}

File: DAIRemoteApp/app/src/main/java/com/example/dairemote_app/utils/ConnectionManager.kt
package com.example.dairemote_app.utils

import android.os.Build
import com.example.dairemote_app.HostSearchCallback
import com.example.dairemote_app.viewmodels.ConnectionViewModel
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.SocketException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.Arrays
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

private const val MAX_RETRIES = 3
private const val INITIAL_RETRY_DELAY_MS = 500L
private const val SOCKET_TIMEOUT_MS = 5000L

class ConnectionManager(serverAddress: String, viewModel: ConnectionViewModel?) {
    private val executorService: ExecutorService
    private lateinit var connectionViewModel: ConnectionViewModel
    private var hostName: String? = null
    private var hostAudioList: String? = null
    private var hostDisplayProfileList: String? = null
    private var hostRequesterResponse: String? = null

    init {
        try {
            setServerAddress(InetAddress.getByName(serverAddress))
            if (viewModel != null) {
                connectionViewModel = viewModel
            }
        } catch (ignored: Exception) {
        }

        this.executorService = Executors.newCachedThreadPool()
    }

    fun setConnectionEstablished(status: Boolean) {
        connectionEstablished.set(status)
    }

    fun getConnectionEstablished(): Boolean {
        return connectionEstablished.get()
    }

    private fun setHostAudioList(list: String?) {
        this.hostAudioList = list
    }

    fun getHostAudioList(): String? {
        return this.hostAudioList
    }

    private fun setHostDisplayProfilesList(list: String?) {
        this.hostDisplayProfileList = list
    }

    fun getHostDisplayProfilesList(): String? {
        return this.hostDisplayProfileList
    }

    private fun setHostRequesterResponse(response: String?) {
        this.hostRequesterResponse = response
    }

    private fun getHostRequesterResponse(): String? {
        return this.hostRequesterResponse
    }

    fun setHostName(name: String?) {
        this.hostName = name
    }

    fun getHostName(): String {
        if (this.hostName != null) {
            return hostName!!.substring("HostName: ".length)
        }
        return ""
    }

    // Wait for server response
    private fun waitForResponse(timeout: Int) {
        try {
            getUDPSocket().soTimeout = timeout
            receiveData()
            getUDPSocket().soTimeout = 75

            setServerResponse(String(receivePacket.data, 0, receivePacket.length))
        } catch (ignored: Exception) {
            setServerResponse("")
        }
    }

    private inline fun runWithRetry(maxRetries: Int, block: (attempt: Int) -> Boolean): Boolean {
        var lastException: Exception? = null

        for (attempt in 0 until maxRetries) {
            try {
                return block(attempt)
            } catch (e: Exception) {
                lastException = e
                if (attempt < maxRetries - 1) {
                    Thread.sleep(INITIAL_RETRY_DELAY_MS * (attempt + 1))
                }
            }
        }

        throw lastException ?: IllegalStateException("No exception recorded")
    }

    fun connectToHost() {
        sendData("Connection requested by ${getDeviceName()}", getInetAddress())
    }

    fun initializeConnection(): Boolean {
        return runWithRetry(MAX_RETRIES) { attempt ->
            try {
                val socketTimeout = SOCKET_TIMEOUT_MS * (attempt + 1)
                getUDPSocket().soTimeout = socketTimeout.toInt()

                connectToHost()
                waitForResponse(socketTimeout.toInt())

                return@runWithRetry when (getServerResponse()?.lowercase()) {
                    "wait" -> {
                        waitForApproval()
                        val approved = getServerResponse().equals("approved", ignoreCase = true)
                        if (approved) {
                            //shutdownHostSearchInBackground()
                            ConnectionMonitor.getInstance(this, connectionViewModel)
                            setConnectionEstablished(true)
                        }
                        return approved
                    }

                    "approved" -> {
                        //shutdownHostSearchInBackground()
                        ConnectionMonitor.getInstance(this, connectionViewModel)
                        setConnectionEstablished(true)
                        true
                    }

                    "declined" -> {
                        resetConnectionManager()
                        false
                    }

                    else -> throw IOException("Invalid server response ${getServerResponse()}")
                }
            } catch (e: Exception) {
                if (attempt == MAX_RETRIES - 1) {
                    resetConnectionManager()
                }
                throw e
            }
        }
    }

    private fun waitForApproval() {
        var approvalTimeout = 0
        setServerResponse("")
        while (getServerResponse().isNullOrEmpty() && approvalTimeout <= 5) {
            waitForResponse(10000)
            approvalTimeout++
        }
    }

    private fun stopExecServices(service: ExecutorService?) {
        if (service != null && !service.isShutdown) {
            service.shutdownNow()
        }
    }

    // Send message to the server
    fun sendHostMessage(msg: String): Boolean {
        if (getConnectionEstablished()) {
            executorService.submit { sendMessage(msg) }
            return true
        }
        return false
    }

    fun sendHostMessage(msg: String, inetAddress: InetAddress): Boolean {
        if (getConnectionEstablished()) {
            executorService.submit { sendMessage(msg, inetAddress) }
            return true
        }
        return false
    }

    private fun sendMessage(message: String) {
        try {
            sendData(message, getInetAddress())
        } catch (ignored: Exception) {
        }
    }

    private fun sendMessage(message: String, inetAddress: InetAddress) {
        try {
            sendData(message, inetAddress)
        } catch (ignored: Exception) {
        }
    }

    private fun hostRequester(
        replyCondition: String?,
        sendMessage: String,
        inetAddress: InetAddress?
    ): Boolean {
        setServerResponse("")
        var broadcastCount = 0
        while (!getServerResponse()!!.startsWith(replyCondition!!)) {
            try {
                sendData(sendMessage, inetAddress)
            } catch (ignored: Exception) {
            }
            broadcastCount += 1
            if (broadcastCount > 5) {
                return false
            } else {
                // Updates serverResponse else times out and throws socket exception
                try {
                    waitForResponse(5000)
                } catch (ignored: Exception) {
                }
            }
        }
        setHostRequesterResponse(getServerResponse())
        return true
    }

    /*    fun requestHostName(): Boolean {
            if (hostRequester("HostName", "HOST Name", getInetAddress())) {
                setHostName(getHostRequesterResponse())
                return true
            }
            return false
        }*/

    // Retrieve audio devices from host
    fun requestHostAudioDevices(): Boolean {
        if (hostRequester("AudioDevices", "AUDIO Devices", getInetAddress())) {
            setHostAudioList(getHostRequesterResponse())
            return true
        }
        return false
    }

    // Retrieve display profiles from host
    fun requestHostDisplayProfiles(): Boolean {
        if (hostRequester("DisplayProfiles", "DISPLAY Profiles", getInetAddress())) {
            setHostDisplayProfilesList(getHostRequesterResponse())
            return true
        }
        return false
    }

    fun resetConnectionManager() {
        stopExecServices(executorService)
        setConnectionEstablished(false)
        setServerResponse(null)
        (null as String?)?.let { setServerAddress(it) }
        (null as InetAddress?)?.let { setServerAddress(it) }
    }

    // Shutdown the connection
    fun shutdown() {
        ConnectionMonitor.getInstance(this, connectionViewModel)?.shutDownHeartbeat()
        sendHostMessage("Shutdown requested")
        try {
            Thread.sleep(75)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
        resetConnectionManager()
        closeUDPSocket()
    }

    fun closeUDPSocket() {
        if (!getUDPSocket().isClosed) {
            getUDPSocket().close()
        }
    }

    companion object {
        private var hostSearchExecService: ExecutorService? = null
        private val connectionEstablished = AtomicBoolean(false)
        private lateinit var serverAddress: String
        private var serverResponse: String? = null
        private lateinit var inetAddress: InetAddress
        private const val SERVER_PORT = 9416
        private var hostHandler: HostSearchCallback? = null

        private var receiveData: ByteArray = ByteArray(1024)
        private var sendData = ByteArray(200)
        private var sendPacket: DatagramPacket? = null
        var receivePacket: DatagramPacket = DatagramPacket(receiveData, receiveData.size)
        private lateinit var udpSocket: DatagramSocket

        init {
            try {
                setUDPSocket(DatagramSocket())
            } catch (e: SocketException) {
                throw RuntimeException(e)
            }
        }

        private var broadcastAddress: InetAddress? = null

        init {
            try {
                broadcastAddress = InetAddress.getByName("255.255.255.255")
            } catch (e: UnknownHostException) {
                throw RuntimeException(e)
            }
        }

        private fun setUDPSocket(socket: DatagramSocket) {
            udpSocket = socket
        }

        fun getUDPSocket(): DatagramSocket {
            if (udpSocket.isClosed) {
                setUDPSocket(DatagramSocket())
            }
            return udpSocket
        }

        fun setServerAddress(address: String) {
            serverAddress = address
        }

        fun setServerAddress(address: InetAddress) {
            inetAddress = address
        }

        fun getServerAddress(): String {
            return if (::serverAddress.isInitialized) serverAddress else ""
        }

        fun getInetAddress(): InetAddress {
            return inetAddress
        }

        fun getPort(): Int {
            return SERVER_PORT
        }

        private fun setServerResponse(response: String?) {
            serverResponse = response
        }

        fun getServerResponse(): String? {
            return serverResponse
        }

        @Throws(IOException::class)
        fun sendData(message: String, address: InetAddress?) {
            sendData = message.toByteArray()
            sendPacket = DatagramPacket(sendData, sendData.size, address, getPort())
            getUDPSocket().send(sendPacket)
        }

        @Throws(IOException::class)
        fun receiveData() {
            // Clear prior data
            Arrays.fill(receiveData, 0.toByte())
            getUDPSocket().receive(receivePacket)
        }

        // Utility to get the device name
        fun getDeviceName(): String {
            val manufacturer = Build.MANUFACTURER
            val model = Build.MODEL
            return if (model.startsWith(manufacturer)) {
                model
            } else {
                "$manufacturer $model"
            }
        }

        // Broadcast to search for hosts in the background
        fun hostSearchInBackground(hostSearchCallback: HostSearchCallback?, message: String) {
            hostHandler = hostSearchCallback
            hostSearchExecService = Executors.newSingleThreadExecutor()
            // Perform the host search
            hostSearchExecService!!.execute { hostSearch(message) }
        }

        fun shutdownHostSearchInBackground() {
            hostSearchExecService!!.shutdownNow()
        }

        // Broadcast to search for hosts
        private fun hostSearch(message: String) {
            val hosts: MutableList<String> = ArrayList()
            try {
                getUDPSocket().broadcast = true
                sendData(message + " " + getDeviceName(), broadcastAddress)
                getUDPSocket().soTimeout =
                    3000 // Millisecond timeout for responses and to break out of loop
                while (true) {
                    try {
                        receiveData() // Blocks until a response is received

                        setServerResponse(String(receivePacket.data).trim { it <= ' ' })
                        val serverIp = receivePacket.address.hostAddress

                        if (getServerResponse()!!.contains("Hello, I'm")) {
                            if (serverIp != null) {
                                hosts.add(serverIp)
                            }
                            getUDPSocket().soTimeout =
                                75 // Reset timeout on host found, otherwise lingers
                            getUDPSocket().broadcast = false
                        }
                    } catch (e: SocketTimeoutException) {
                        // Stop listening for responses, occurs on timeout
                        break
                    }
                }

                // Reset server response, just being used for broadcast search here
                // Needs to be empty for proceeding functions
                setServerResponse(null)

                if (hosts.isNotEmpty()) {
                    hostHandler!!.onHostFound(hosts)
                } else {
                    hostHandler!!.onError("No hosts found")
                }
            } catch (ignored: SocketException) {
            } catch (e: UnknownHostException) {
                throw RuntimeException(e)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }
}